MTC Enhancements March 2020

landonreed opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

MTC enhancements re #544. 
- Reverse version dropdown sort
- Do not allow selecting of merged feeds in merge feeds dropdown
- Add retrieval method icon to dropdown items/version summary

evansiroky commented on lib/manager/components/version/FeedVersionDetails.js:
```diff
@@ -51,12 +52,30 @@ export default class FeedVersionDetails extends Component<Props> {
+        ? '(Cannot re-merge feed)'
+        : v.name
+    const disabled = v.retrievalMethod === 'SERVICE_PERIOD_MERGE' ||
+      v.id === activeVersion.id
```
Nitpick: might be better to use if/else if blocks here to make things more readable and avoid duplicate code.

landonreed replied:
Something like:
```
let name = v.name
    let disabled = false
    if (v.id === activeVersion.id) {
      name = '(Cannot merge with self)'
      disabled = true
    }
    if (v.retrievalMethod === 'SERVICE_PERIOD_MERGE') {
      name = '(Cannot re-merge feed)'
      disabled = true
    }
```
?

binh-dam-ibigroup replied:
Actually does it make sense at all to display versions that cannot be merged?

landonreed replied:
I think it works either way. I'm inclined to just disable them so they don't appear to be missing to the user without explanation.

evansiroky commented on lib/manager/components/version/VersionRetrievalBadge.js:
```diff
@@ -0,0 +1,40 @@
+// @flow
+
+import Icon from '@conveyal/woonerf/components/icon'
+import * as React from 'react'
```
I don't think we need to import everything from React here. Seems like `React, {Component}` should suffice.

landonreed replied:
I think that was leftover from when I was trying to specify a flow type of React.Node.

evansiroky commented on lib/manager/components/version/VersionRetrievalBadge.js:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+export default class VersionRetrievalBadge extends React.Component<Props> {
```
This could be just a function, no need to make it a full component.

evansiroky commented on lib/manager/components/version/VersionSelectorDropdown.js:
```diff
@@ -0,0 +1,77 @@
+  )
+}
+
+export default class VersionSelectorDropdown extends React.Component<Props> {
```
This could be just a function, no need to make it a full component.

evansiroky commented on lib/manager/components/version/VersionSelectorDropdown.js:
```diff
@@ -0,0 +1,77 @@
+  versions: ?Array<FeedVersion>
+}
+
+const defaultItemFormatter = (version: FeedVersion, index: number, activeVersion: FeedVersion) => {
```
Make this into a functional component. Also, is the index argument needed/used?

binh-dam-ibigroup commented on lib/manager/actions/status.js:
```diff
@@ -134,6 +136,51 @@ export async function fetchAppInfo () {
+    const skippedRecordsForTables = {}
+    result.skippedIds.forEach(id => {
+      const table = id.split(':')[0]
+      if (skippedRecordsForTables[table]) {
```
Suggestion: 
`const skippedRecords = skippedRecordsForTables[table]
skippedRecordsForTables[table] = skippedRecords ? 1 : skippedRecords + 1 
`

landonreed replied:
I think your ternary is backwards?

binh-dam-ibigroup replied:
Probably 🤔

landonreed replied:
How about:
```
skippedRecordsForTables[table] = (skippedRecordsForTables[table] || 0) + 1
```
That doesn't seem too clever, but is nice and concise.

binh-dam-ibigroup replied:
That works too.

binh-dam-ibigroup commented on lib/manager/actions/status.js:
```diff
@@ -134,6 +136,51 @@ export async function fetchAppInfo () {
+  }
+}
+
 /* eslint-disable complexity */
```
Can you remove `/* eslint-disable complexity */` now?

landonreed replied:
I'm not sure if this is needed anywhere anymore. When I remove these `yarn lint` runs fine. @evansiroky can you weigh in on whether these are still needed?

binh-dam-ibigroup commented on lib/manager/actions/status.js:
```diff
@@ -203,60 +250,21 @@ export function handleFinishedJob (job: ServerJob) {
-                : `Merge was completed successfully. A new version will be processed/validated containing the resulting feed.`,
-              detail: details.join('\n')
-            }))
+            const modalContent = getMergeFeedModalContent(result)
```
Inline this.

binh-dam-ibigroup commented on lib/manager/components/version/VersionSelectorDropdown.js:
```diff
@@ -0,0 +1,77 @@
+      header,
+      itemFormatter,
+      title,
+      version,
```
Suggestion: also rename `version` to `activeVersion`.

landonreed replied:
Leaving it named version conforms with the name of the prop in other version components. I do somewhat agree with you on this item, but I think it should be handled in a separate PR and deserves a little more consideration on whether we should make the change.

landonreed commented:
@evansiroky, I have addressed your comments. @binh-dam-ibigroup, I addressed one comment, but skipped the eslint disable one based on our discussion on IBI chat.  Also, I'm not keen on changing the prop name unless we did it wholesale throughout other components, which seems a bit out of scope.
